### PR TITLE
test(skills): add Hypothesis property tests for PowerPoint modules

### LINKED ISSUE #1013: feat(powerpoint): add Hypothesis property tests for priority modules
## Summary

Add Hypothesis property-based testing to the PowerPoint skill's priority modules. The codebase currently has ~300+ deterministic unit tests with well-formed inputs but zero property-based or fuzz tests. Hypothesis will generate structured adversarial inputs (malformed YAML dicts, edge-case values, type-confused inputs) that exercise business logic with invalid data — the primary gap in the current test suite.

## Context

This is **Phase 1** of the Python Security Testing & Fuzzing Initiative. Hypothesis was selected over ClusterFuzzLite/Atheris as the primary testing approach because:
- The codebase's primary input is structured YAML dictionaries — Hypothesis generates valid structured inputs natively while Atheris wastes cycles on syntactically invalid raw bytes
- Zero infrastructure overhead (no Docker, storage repos, or PAT tokens)
- Native pytest integration with the existing test suite
- Automatic test case shrinking for minimal reproducing examples

### Dependencies

Add to `.github/skills/experimental/powerpoint/pyproject.toml`:

```toml
[dependency-groups]
dev = [
    "hypothesis>=6.100",
    # ... existing dev deps
]
```

### Priority Test Targets

**1. validate_slides.py / validate_deck.py — Input validation robustness**

```python
from hypothesis import given, settings
import hypothesis.strategies as st

yaml_values = st.recursive(
    st.none() | st.booleans() | st.integers() | st.floats(allow_nan=False) | st.text(),
    lambda children: st.lists(children) | st.dictionaries(st.text(), children),
    max_leaves=50,
)

@given(data=yaml_values)
@settings(max_examples=500, deadline=None)
def test_validate_slides_handles_arbitrary_yaml(data):
    """Property: validate functions never raise unhandled exceptions on arbitrary input."""
    try:
        validate_slides(data)
    except (ValueError, TypeError, KeyError):
        pass  # Expected rejections
```

**2. build_deck.py — Element builder dispatch robustness**

```python
slide_element = st.fixed_dictionaries({
    "type": st.sampled_from(["text", "image", "table", "chart", "shape", "group", "connector"]),
}, optional={
    "x": st.floats(min_value=-100, max_value=100),
    "y": st.floats(min_value=-100, max_value=100),
    "width": st.floats(min_value=0, max_value=100),
    "height": st.floats(min_value=0, max_value=100),
    "text": st.text(max_size=1000),
    "rows": st.integers(min_value=0, max_value=50),
    "columns": st.integers(min_value=0, max_value=50),
})

@given(elements=st.lists(slide_element, max_size=20))
@settings(max_examples=300, deadline=None)
def test_build_slide_handles_arbitrary_elements(elements):
    """Property: build_slide handles arbitrary element definitions without crashing."""
    ...
```

**3. pptx_colors.py — Hex parsing edge cases**

```python
@given(hex_str=st.text(alphabet="0123456789abcdefABCDEF", min_size=0, max_size=20))
def test_resolve_color_hex_robustness(hex_str):
    """Property: resolve_color handles any hex-like string."""
    try:
        resolve_color(hex_str)
    except (ValueError, TypeError):
        pass
```

**4. pptx_tables.py — Merge bounds validation**

```python
@given(
    rows=st.integers(min_value=1, max_value=20),
    cols=st.integers(min_value=1, max_value=20),
    merge_right=st.integers(min_value=-5, max_value=25),
    merge_down=st.integers(min_value=-5, max_value=25),
)
def test_table_merge_bounds(rows, cols, merge_right, merge_down):
    """Property: table merge operations handle out-of-bounds gracefully."""
    ...
```

### File Organization

New test files follow the existing pattern under `.github/skills/experimental/powerpoint/tests/`:
- `test_fuzz_validate_slides.py`
- `test_fuzz_build_deck.py`
- `test_fuzz_pptx_colors.py`
- `test_fuzz_pptx_tables.py`
- `test_fuzz_extract_content.py` (for XML parsing and blob extraction)

### task-researcher
- Review each priority module's public API and input types
- Identify which functions accept external/untrusted input
- Catalog existing edge cases already covered by deterministic tests
- Determine appropriate Hypothesis strategy configurations for each input type

### task-planner
- Define property invariants for each target function (what should always be true)
- Design custom Hypothesis strategies matching the YAML slide definition schema
- Plan test file organization and naming conventions
- Set `max_examples` and `deadline` settings appropriate for CI runtime

### task-implementor
- Add `hypothesis>=6.100` to `pyproject.toml` dev dependencies
- Implement property tests for all 4 priority targets (minimum)
- Extend to `extract_content.py` and `pptx_text.py` as stretch goals
- Verify all tests pass in CI and Hypothesis database is persisted
- Document property test conventions for future contributors

## Acceptance Criteria

- [ ] `hypothesis>=6.100` is added to `pyproject.toml` dev dependencies
- [ ] Property tests exist for `validate_slides.py`, `build_deck.py`, `pptx_colors.py`, and `pptx_tables.py` (minimum 4 modules)
- [ ] Each property test defines clear invariants (not just "doesn't crash")
- [ ] Custom Hypothesis strategies generate inputs matching the YAML slide definition schema
- [ ] All property tests pass in CI with `max_examples >= 200`
- [ ] Hypothesis database directory is configured for persistence across runs
- [ ] No unhandled exceptions are discovered (or discovered exceptions are fixed or documented)
- [ ] Test execution completes within reasonable CI time limits (< 5 minutes for property tests)

### PR DESCRIPTION
**Pull request by WilliamBerryiii**
## Description

Added Hypothesis property-based testing to the PowerPoint skill, covering four core modules with 18 property tests. A shared strategies module and CI/dev Hypothesis profiles provide the testing infrastructure. Property testing uncovered a numeric sort bug in `discover_images`, which was fixed in this PR.

### Property-Based Tests

Eighteen property tests exercised invariants, round-trip behavior, boundary conditions, and idempotency across four production modules.

- Added **test_fuzz_pptx_colors.py** with six tests: `resolve_color` always-returns-dict, contains-rgb-or-theme, hex-idempotent; `hex_brightness` result 0–255; `rgb_to_hex` output format; and color round-trip preserving color type through shape fill.
- Added **test_fuzz_pptx_tables.py** with four tests: table round-trip preserving row count, column count, and cell text through add/extract; merge boundary raising `IndexError` for out-of-bounds spans.
- Added **test_fuzz_build_deck.py** with three tests: `discover_slides` output-sorted and only-matching-dirs invariants; `get_slide_layout` always-returns-layout for arbitrary names.
- Added **test_fuzz_validate_slides.py** with five tests: `discover_images` output-is-sorted and all-paths-match-pattern; `max_severity` ordering-monotonic, empty-returns-none, and result-in-valid-set.

### Shared Test Infrastructure

- Added *strategies.py* with reusable Hypothesis strategies for hex colors, theme references, table elements, positions, issues, and severity results. The `table_element` composite strategy restricts text to XML-safe Unicode categories.
- Updated *conftest.py* with `ci` and `dev` Hypothesis profiles (200 vs 50 max_examples, derandomize in CI, suppressed `too_slow` health check), plus `make_blank_presentation()` and `make_blank_slide()` helper functions as plain functions for use inside `@given`-decorated tests.

### Bug Fix

- Fixed **discover_images** in *validate_slides.py* to sort by numeric slide number instead of lexicographic filesystem order. Previous behavior sorted `slide-10` before `slide-2`.

### Configuration

- Added `hypothesis>=6.100` dev dependency and `hypothesis` pytest marker to *pyproject.toml*.
- Configured `[tool.coverage.run]` source=["scripts"] and `[tool.coverage.report]` fail_under=85 with show_missing.
- Added `**/.hypothesis/` to *.gitignore* for Hypothesis example database directories.

## Related Issue(s)

Closes #1013

## Type of Change

Select all that apply:

**Code & Documentation:**

* [ ] Bug fix (non-breaking change fixing an issue)
* [x] New feature (non-breaking change adding functionality)
* [ ] Breaking change (fix or feature causing existing functionality to change)
* [ ] Documentation update

**Infrastructure & Configuration:**

* [ ] GitHub Actions workflow
* [ ] Linting configuration (markdown, PowerShell, etc.)
* [ ] Security configuration
* [ ] DevContainer configuration
* [x] Dependency update

**AI Artifacts:**

* [ ] Reviewed contribution with `prompt-builder` agent and addressed all feedback
* [ ] Copilot instructions (`.github/instructions/*.instructions.md`)
* [ ] Copilot prompt (`.github/prompts/*.prompt.md`)
* [ ] Copilot agent (`.github/agents/*.agent.md`)
* [ ] Copilot skill (`.github/skills/*/SKILL.md`)

> Note for AI Artifact Contributors:
>
> * Agents: Research, indexing/referencing other project (using standard VS Code GitHub Copilot/MCP tools), planning, and general implementation agents likely already exist. Review `.github/agents/` before creating new ones.
> * Skills: Must include both bash and PowerShell scripts. See [Skills](../docs/contributing/skills.md).
> * Model Versions: Only contributions targeting the **latest Anthropic and OpenAI models** will be accepted. Older model versions (e.g., GPT-3.5, Claude 3) will be rejected.
> * See [Agents Not Accepted](../docs/contributing/custom-agents.md#agents-not-accepted) and [Model Version Requirements](../docs/contributing/ai-artifacts-common.md#model-version-requirements).

**Other:**

* [x] Script/automation (`.ps1`, `.sh`, `.py`)
* [ ] Other (please describe):

## Testing

- 587 tests passed after merging with origin/main (up from 549 on the base branch).
- 18 new Hypothesis property tests verified invariants, round-trip behavior, boundary conditions, and idempotency across `pptx_colors`, `pptx_tables`, `build_deck`, and `validate_slides`.
- Coverage measured at 90%, exceeding the 85% `fail_under` threshold.
- Ruff linting passed with no issues.

### Required Checks

* [ ] Documentation is updated (if applicable)
* [x] Files follow existing naming conventions
* [ ] Changes are backwards compatible (if applicable)
* [x] Tests added for new functionality (if applicable)

### AI Artifact Contributions

* [ ] Used `/prompt-analyze` to review contribution
* [ ] Addressed all feedback from `prompt-builder` review
* [ ] Verified contribution follows common standards and type-specific requirements

### Required Automated Checks

The following validation commands must pass before merging:

* [x] Markdown linting: `npm run lint:md`
* [x] Spell checking: `npm run spell-check`
* [x] Frontmatter validation: `npm run lint:frontmatter`
* [x] Skill structure validation: `npm run validate:skills`
* [x] Link validation: `npm run lint:md-links`
* [x] PowerShell analysis: `npm run lint:ps`
* [x] Plugin freshness: `npm run plugin:generate`

## Security Considerations

* [x] This PR does not contain any sensitive or NDA information
* [x] Any new dependencies have been reviewed for security issues
* [x] Security-related scripts follow the principle of least privilege

## Additional Notes

- The `hypothesis` package (6.151.9) is a well-established property-based testing library with no known security vulnerabilities. Its only transitive dependency is `sortedcontainers`.
- Hypothesis profiles are automatically selected based on the `CI` environment variable, with the `ci` profile using derandomized inputs and no persistent database.
- The `discover_images` bug fix addressed a numeric vs. lexicographic sort issue that would have caused incorrect slide ordering for decks with 10+ slides.

### CODE REVIEW

File: .github/skills/experimental/powerpoint/tests/test_fuzz_validate_slides.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+        with tempfile.TemporaryDirectory() as tmp:
+            tmp_path = Path(tmp)
+            for n in slide_nums:
+                (tmp_path / f"slide-{n}.jpg").write_bytes(b"fake")
+            for name in noise_names:
+                path = tmp_path / name
+                if not path.exists():
+                    try:
+                        path.write_bytes(b"noise")
+                    except OSError:
```
[the remark this answers is not in the document]
Comment (WilliamBerryiii) [reply .github/skills/experimental/powerpoint/tests/test_fuzz_validate_slides.py:59]: Fixed in a314295f — added an explanatory comment clarifying that Hypothesis-generated filenames with invalid characters are expected to fail, so the `OSError` is intentionally swallowed.

File: .github/skills/experimental/powerpoint/tests/strategies.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
+"""Shared Hypothesis strategies for PowerPoint skill property tests."""
+
+from hypothesis import strategies as st
+
+# Theme names matching THEME_COLOR_MAP keys in pptx_colors.py
+THEME_NAMES = [
```
Comment (auyidi1) [inline .github/skills/experimental/powerpoint/tests/strategies.py]: These theme names appear to be hardcoded instead of using the names in `pptx_colors.py`. Why not use the actual names instead:
`THEME_NAMES = list(THEME_COLOR_MAP.keys())`

File: .github/skills/experimental/powerpoint/tests/test_fuzz_pptx_colors.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+        from pptx.util import Inches
+
+        shape = slide.shapes.add_shape(
+            MSO_SHAPE.RECTANGLE, Inches(1), Inches(1), Inches(2), Inches(2)
+        )
+        shape.fill.solid()
+        color_spec = {"rgb": RGBColor(r, g, b)}
+        apply_color_spec(shape.fill.fore_color, color_spec)
+        extracted = extract_color(shape.fill.fore_color)
+        assert isinstance(extracted, str)
```
Comment (auyidi1) [inline .github/skills/experimental/powerpoint/tests/test_fuzz_pptx_colors.py]: This assertion only checks type, not value. The extracted hex should exactly match the input. Assert on extracted hex matches f"#{r:02X}{g:02X}{b:02X}" for RGB inputs

File: .github/skills/experimental/powerpoint/tests/conftest.py [hunk 4]
```diff
@@ -81,3 +100,18 @@ def sample_image_path(tmp_path):
     img = tmp_path / "test.png"
     img.write_bytes(_minimal_png_bytes())
     return img
+
+
+def make_blank_presentation():
```
Comment (auyidi1) [inline .github/skills/experimental/powerpoint/tests/conftest.py]: Duplicates `blank_presentation` fixture from above with no explanation:
`def blank_presentation():
    """Fresh Presentation with standard widescreen dimensions."""
    prs = Presentation()
    prs.slide_width = Inches(13.333)
    prs.slide_height = Inches(7.5)
    return prs`

File: .github/skills/experimental/powerpoint/tests/test_fuzz_pptx_colors.py [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+
+    @settings(deadline=None)
+    @given(
+        r=st.integers(min_value=0, max_value=255),
+        g=st.integers(min_value=0, max_value=255),
+        b=st.integers(min_value=0, max_value=255),
+    )
+    def test_apply_extract_preserves_color_type(self, r, g, b):
+        slide = make_blank_slide()
+        from pptx.enum.shapes import MSO_SHAPE
```
Comment (auyidi1) [inline .github/skills/experimental/powerpoint/tests/test_fuzz_pptx_colors.py]: There are imports here inside the test body for `MSO_SHAPE` and `Inches`. Move these imports to the module-level above.

File: .github/skills/experimental/powerpoint/tests/conftest.py (hunk 4 above)
Comment (WilliamBerryiii) [reply .github/skills/experimental/powerpoint/tests/conftest.py]: Refactored — the `blank_presentation` fixture now delegates to `make_blank_presentation()`, eliminating the code duplication. The plain functions exist because `@given`-decorated Hypothesis tests can't use pytest fixtures; added a section comment explaining this.

File: .github/skills/experimental/powerpoint/tests/test_fuzz_pptx_colors.py (hunk 3 above)
Comment (WilliamBerryiii) [reply .github/skills/experimental/powerpoint/tests/test_fuzz_pptx_colors.py]: Agreed — strengthened the assertion to compare the exact hex value: `assert extracted == f"#{r:02X}{g:02X}{b:02X}"` instead of just checking type and prefix.

File: .github/skills/experimental/powerpoint/tests/strategies.py (hunk 2 above)
Comment (WilliamBerryiii) [reply .github/skills/experimental/powerpoint/tests/strategies.py]: Good call — replaced the hardcoded list with `THEME_NAMES = list(THEME_COLOR_MAP.keys())` imported from `pptx_colors`. This keeps the test strategies in sync with production automatically.

File: .github/skills/experimental/powerpoint/tests/test_fuzz_pptx_colors.py (hunk 5 above)
Comment (WilliamBerryiii) [reply .github/skills/experimental/powerpoint/tests/test_fuzz_pptx_colors.py]: Moved `MSO_SHAPE` and `Inches` imports to module-level.

File: .github/skills/experimental/powerpoint/tests/strategies.py [hunk 6]
```diff
@@ -0,0 +1,80 @@
+"""Shared Hypothesis strategies for PowerPoint skill property tests."""
```
Comment (auyidi1) [inline .github/skills/experimental/powerpoint/tests/strategies.py:3]: Are we missing the copyright headers in the new files?

File: .github/skills/experimental/powerpoint/tests/test_fuzz_validate_slides.py [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+            for _, p in result:
+                assert IMAGE_PATTERN.match(p.name)
+
+
+@pytest.mark.hypothesis
+class TestFuzzMaxSeverity:
+    """Property tests for max_severity."""
+
+    @given(data=severity_results)
+    def test_ordering_monotonic(self, data):
```
Comment (auyidi1) [inline .github/skills/experimental/powerpoint/tests/test_fuzz_validate_slides.py]: Medium: `test_ordering_monotonic` only proves a trivial property

`The test injects an issue with `"severity": "error"` and then asserts the result equals `"error"`. This is mathematically guaranteed by the `max_severity` implementation (L246-261 in `validate_deck.py`), making the `>=` comparison at L91 redundant, as both `max_severity(augmented)` and `"error"` have rank 3 always.

A stronger property test would parametrize the injected severity (info, warning, error) and verify the monotonicity invariant: $\text{rank}(\text{result with injection}) \geq \max(\text{rank}(\text{baseline}), \text{rank}(\text{injected}))$. This would catch bugs where, for example, adding a `"warning"` to an `"error"` baseline accidentally downgraded the result.

File: .github/skills/experimental/powerpoint/tests/test_fuzz_pptx_colors.py [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+        assert isinstance(result, str)
+        assert re.fullmatch(r"#[0-9A-Fa-f]{6}", result)
+
+
+@pytest.mark.hypothesis
+class TestFuzzColorRoundTrip:
+    """Property tests for apply/extract color round-trip."""
+
+    @settings(deadline=None)
+    @given(
```
Comment (auyidi1) [inline .github/skills/experimental/powerpoint/tests/test_fuzz_pptx_colors.py:77]: Medium: `test_apply_extract_preserves_color_type` creates 200 Presentation objects in CI

Each `@given` invocation calls `make_blank_slide()`, which calls `make_blank_presentation()`, which instantiates a full `Presentation()` object and adds a slide. With `max_examples=200` in CI profile (`conftest.py` L16), this creates 200 independent Presentation objects. The `Presentation()` constructor parses the full default template PPTX from python-pptx's package data each time. While this works, it is avoidably slow and likely the reason `suppress_health_check=[HealthCheck.too_slow]` was needed (`conftest.py` L21).

Consider creating the Presentation once and adding shapes to a shared slide, or using `@settings(max_examples=50)` for this specific heavy test.

File: .github/skills/experimental/powerpoint/tests/test_fuzz_validate_slides.py [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+                }
+            ],
+            "deck_issues": data.get("deck_issues", []),
+        }
+        assert max_severity(augmented) == "error"
+        # Adding error never decreases severity
+        severity_rank = {"none": 0, "info": 1, "warning": 2, "error": 3}
+        assert severity_rank[max_severity(augmented)] >= severity_rank[baseline]
+
+    def test_empty_returns_none(self):
```
Comment (auyidi1) [inline .github/skills/experimental/powerpoint/tests/test_fuzz_validate_slides.py]: Low: `test_empty_returns_none` name suggests Python `None`, returns string `"none"`

`max_severity({"slides": []})` returns the string `"none"`, not Python `None`. The test name `test_empty_returns_none` could mislead readers into thinking the function returns `None`. A name like `test_empty_returns_none_string` would be clearer.

File: .github/skills/experimental/powerpoint/tests/strategies.py (hunk 6 above)
Comment (WilliamBerryiii) [reply .github/skills/experimental/powerpoint/tests/strategies.py:3]: if we are ... then the build job is broken which appears to be the case ... filing an issue.

Comment (WilliamBerryiii) [reply .github/skills/experimental/powerpoint/tests/strategies.py:3]: oh ... the build IS breaking ... that's good. when I addressed the prior feedback with the file extraction it missed the header, but the build correctly broke!

File: .github/skills/experimental/powerpoint/tests/test_fuzz_validate_slides.py (hunk 9 above)
Comment (WilliamBerryiii) [reply .github/skills/experimental/powerpoint/tests/test_fuzz_validate_slides.py]: Fixed in 3df86b1b — renamed to `test_empty_returns_none_string` to clarify that `max_severity` returns the string `"none"`, not Python `None`.

File: .github/skills/experimental/powerpoint/tests/test_fuzz_pptx_colors.py (hunk 8 above)
Comment (WilliamBerryiii) [reply .github/skills/experimental/powerpoint/tests/test_fuzz_pptx_colors.py:77]: Fixed in 3df86b1b — added `max_examples=50` to the `@settings` decorator for `test_apply_extract_preserves_color_type`, overriding the CI profile's 200. This reduces Presentation object instantiation by 75% for this specific heavy test while keeping the broader CI profile at 200 for the lighter tests.

File: .github/skills/experimental/powerpoint/tests/test_fuzz_validate_slides.py (hunk 7 above)
Comment (WilliamBerryiii) [reply .github/skills/experimental/powerpoint/tests/test_fuzz_validate_slides.py]: Fixed in 3df86b1b — `test_ordering_monotonic` now parametrizes the injected severity across `["info", "warning", "error"]` via `st.sampled_from` and verifies two properties:
1. `rank(result) >= rank(baseline)` — adding an issue never decreases severity.
2. `rank(result) >= rank(injected)` — the result is at least as severe as the injected severity.

This catches the class of bugs described (e.g., adding `"warning"` to an `"error"` baseline accidentally downgrading the result).

File: .github/skills/experimental/powerpoint/tests/strategies.py (hunk 6 above)
Comment (WilliamBerryiii) [reply .github/skills/experimental/powerpoint/tests/strategies.py:3]: Fixed in 3df86b1b — added `# Copyright (c) Microsoft Corporation.` and `# SPDX-License-Identifier: MIT` headers to all five new files: `strategies.py`, `test_fuzz_build_deck.py`, `test_fuzz_pptx_colors.py`, `test_fuzz_pptx_tables.py`, and `test_fuzz_validate_slides.py`. Copyright validation now passes at 100% compliance (146/146 files).

**Comment by WilliamBerryiii** [comment]
### npm audit fix: `flatted` 3.4.1 → 3.4.2

`npm audit` flagged a **high-severity Prototype Pollution** vulnerability in `flatted <=3.4.1` ([GHSA-rf6f-7fwh-wjgh](https://github.com/advisories/GHSA-rf6f-7fwh-wjgh)).

**Dependency chain:** `cspell@9.7.0` → `flatted@3.4.1`

**Fix:** Bumped `flatted` to `3.4.2` via `npm audit fix`. Only `package-lock.json` changed (3 insertions, 3 deletions). `npm audit` now reports 0 vulnerabilities and `cspell` spell-check passes cleanly.

Commit: `9c482e1c` — `fix(build): bump flatted 3.4.1 to 3.4.2 for prototype pollution fix`
